compiler/rustc_trait_selection/src/solve/eval_ctxt/canonical/dedup_solver/constraint_walker.rs

```diff
@@ -0,0 +1,130 @@
+use rustc_data_structures::fx::FxIndexSet;
+use rustc_infer::infer::InferCtxt;
+use rustc_middle::ty;
+use rustc_middle::ty::{Const, Region, Ty, TyCtxt, TypeFoldable, TypeFolder, TypeSuperFoldable};
+
+pub struct DedupWalker<'me, 'tcx> {
+    infcx: &'me InferCtxt<'tcx>,
+    var_indexer: &'me mut DedupableIndexer<'tcx>,
+    max_nameable_universe: ty::UniverseIndex,
+
+    vars_present: Vec<usize>,
+}
+pub struct DedupableIndexer<'tcx> {
+    vars: FxIndexSet<DedupVar<'tcx>>,
+    pub unremovable_vars: FxIndexSet<usize>,
+}
+#[derive(Hash, PartialEq, Eq)]
+enum DedupVar<'tcx> {
+    Region(Region<'tcx>),
+    Ty(Ty<'tcx>),
+    Const(Const<'tcx>),
+}
```

Comment:
why this instead of using the existing `GenericArg`?

Reply:
Ohhhh yeah you're right, they are just the same thing. I'll fix that